src/main.js

```diff
@@ -292,6 +295,252 @@ const mixin = {
             var store = tx.objectStore("channel_groups");
             store.delete(groupName);
         },
+        async getLocalPlaylist(playlistId) {
+            var tx = window.db.transaction("playlists", "readonly");
+            var store = tx.objectStore("playlists");
+            const req = store.openCursor(playlistId);
+            let playlist = null;
+            req.onsuccess = e => {
+                playlist = e.target.result.value;
+            };
+            while (playlist == null) {
+                await this.delay(10);
+            }
+            playlist.videos = JSON.parse(playlist.videoIds).length;
+            return playlist;
+        },
+        createOrUpdateLocalPlaylist(playlist) {
+            var tx = window.db.transaction("playlists", "readwrite");
+            var store = tx.objectStore("playlists");
+            store.put(playlist);
+        },
+        // needs to handle both, streamInfo items and streams items
+        createLocalPlaylistVideo(videoId, videoInfo) {
+            if (videoInfo === undefined || videoId === null || videoInfo?.error) return;
+
+            var tx = window.db.transaction("playlist_videos", "readwrite");
+            var store = tx.objectStore("playlist_videos");
+            const video = {
+                videoId: videoId,
+                title: videoInfo.title,
+                type: "stream",
+                shortDescription: videoInfo.shortDescription ?? videoInfo.description,
+                url: `/watch?v=${videoId}`,
+                thumbnail: videoInfo.thumbnail ?? videoInfo.thumbnailUrl,
+                uploaderVerified: videoInfo.uploaderVerified,
+                duration: videoInfo.duration,
+                uploaderAvatar: videoInfo.uploaderAvatar,
+                uploaderUrl: videoInfo.uploaderUrl,
+                uploaderName: videoInfo.uploaderName ?? videoInfo.uploader,
+            };
+            store.put(video);
+        },
+        async getLocalPlaylistVideo(videoId) {
+            var tx = window.db.transaction("playlist_videos", "readonly");
+            var store = tx.objectStore("playlist_videos");
+            const req = store.openCursor(videoId);
+            let video = null;
+            req.onsuccess = e => {
+                video = e.target.result.value;
+            };
+            while (video == null) {
+                await this.delay(10);
+            }
+            return video;
```

Comment:
I really don't like this Busy Waiting approach. Why is this necessary?

Reply:
Because IndexedDB doesn't support `Promise` and the async API. If we didn't do it that way, we would need to do a lot more changes on each playlist component to make it work with adding items one by one too. If you have a better idea to wait until it's done, instead of using short delays, that might be worth it, but I don't think we can avoid it without causing new issues.

Reply:
Since the function is async anyways you could use something like this:
```js
async getLocalPlaylistVideo(videoId) {
  return new Promise((resolve, reject) => {
    [...]
    req.onsuccess = e => {
      resolve(e.target.result.value);
    }
  })
}
```

Reply:
Works like a charm, thanks!